crates/matrix-sdk-crypto/src/types/mod.rs

```diff
@@ -27,6 +27,145 @@ mod cross_signing_key;
+        }
+    }
+}
+
+impl From<Ed25519Signature> for Signature {
+    fn from(signature: Ed25519Signature) -> Self {
+        Self::Ed25519(signature)
+    }
+}
+
+/// Signatures for a signed object.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Signatures(BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, Signature>>);
+
+impl Signatures {
+    /// Create a new, empty, signatures collection.
+    pub fn new() -> Self {
+        Signatures(Default::default())
+    }
+
+    /// Add the given signature from the given signer and the given key_id to
+    /// the collection.
+    pub fn add_signature(
+        &mut self,
+        signer: OwnedUserId,
+        key_id: OwnedDeviceKeyId,
+        signature: Ed25519Signature,
+    ) -> Option<Signature> {
+        self.0.entry(signer).or_insert_with(Default::default).insert(key_id, signature.into())
+    }
+
+    /// Try to find an Ed25519 signature from the given signer with the given
+    /// key id.
+    pub fn get_signature(&self, signer: &UserId, key_id: &DeviceKeyId) -> Option<Ed25519Signature> {
+        self.get(signer)?.get(key_id)?.ed25519()
+    }
+
+    /// Get the map of signatures that belong to the given user.
+    pub fn get(&self, signer: &UserId) -> Option<&BTreeMap<OwnedDeviceKeyId, Signature>> {
+        self.0.get(signer)
+    }
+}
+
+impl Default for Signatures {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl IntoIterator for Signatures {
+    type Item = (OwnedUserId, BTreeMap<OwnedDeviceKeyId, Signature>);
+
+    type IntoIter =
+        std::collections::btree_map::IntoIter<OwnedUserId, BTreeMap<OwnedDeviceKeyId, Signature>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<'de> Deserialize<'de> for Signatures {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let map: BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, String>> =
+            serde::Deserialize::deserialize(deserializer)?;
+
+        Ok(Signatures(
+            map.into_iter()
+                .map(|(u, m)| {
+                    Ok((
+                        u,
+                        m.into_iter()
+                            .map(|(d, s)| {
+                                let algorithm = d.algorithm();
+                                Ok((
+                                    d,
+                                    if algorithm == DeviceKeyAlgorithm::Ed25519 {
+                                        Ed25519Signature::from_base64(&s)
+                                            .map_err(serde::de::Error::custom)?
+                                            .into()
+                                    } else {
+                                        Signature::Other(s)
+                                    },
+                                ))
+                            })
+                            .collect::<Result<BTreeMap<_, _>, _>>()?,
```

Comment:
I think this would be easier to read with this value being a local binding first (and maybe with `u`, `m` getting more obvious names).

Reply:
Hopefully https://github.com/matrix-org/matrix-rust-sdk/pull/716/commits/2f3eccc257b8537a299924dcb0d0b1795c82f748 makes this clearer.